pkg/kube/krt/map.go

```diff
@@ -0,0 +1,163 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"fmt"
+
+	"istio.io/istio/pkg/ptr"
+	"istio.io/istio/pkg/slices"
+)
+
+// MapCollection is just a light facade on top of another collection
+// that uses a map function to trasform T into K
+type mapCollection[T any, U any] struct {
+	collectionName string
+	id             collectionUID
+	collection     internalCollection[T]
+	mapFunc        func(T) U
+}
+
+var _ Collection[int] = &mapCollection[bool, int]{}
+
+// nolint: unused // (not true, used in func declared to implement an interface)
+type mappedIndexer[T any, U any] struct {
+	indexer indexer[T]
+	mapFunc func(T) U
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (m *mappedIndexer[T, U]) Lookup(k string) []U {
+	var res []U
+	for _, obj := range m.indexer.Lookup(k) {
+		res = append(res, m.mapFunc(obj))
+	}
+	return res
+}
+
+func (m *mapCollection[T, U]) GetKey(k string) *U {
+	if obj := m.collection.GetKey(k); obj != nil {
+		return ptr.Of(m.mapFunc(*obj))
+	}
+	return nil
+}
+
+func (m *mapCollection[T, U]) List() []U {
+	var res []U
+	for _, obj := range m.collection.List() {
+		res = append(res, m.mapFunc(obj))
+	}
+	return res
```

Comment:
same as above, probably best to allocate `res` with the exact amount necessary

Reply:
Good call outs; I wrote this in the nested collection context where we can't always pre-allocate, so missed some of these easy optimizations